### PR TITLE
Regexp tidy up

### PR DESCRIPTION
**Pull request by jimmidyson**
Cleaned up some areas where regexps were either messy or unnecessary.

### CODE REVIEW

File: container/docker/factory.go [hunk 1]
```diff
@@ -172,9 +172,12 @@ func (self *dockerFactory) DebugInfo() map[string][]string {
 	return map[string][]string{}
 }
 
+var (
+	version_regexp_string = "(\\d+)\\.(\\d+)\\.(\\d+)"
+	version_re            = regexp.MustCompile(version_regexp_string)
```
Comment (timstclair) [inline container/docker/factory.go]: What's the point of moving this out of the function? Personally I find it more readable if I can see the regexp right where it's used (and I don't think this should be used outside that function).

File: container/docker/factory.go [hunk 2]
```diff
@@ -172,9 +172,12 @@ func (self *dockerFactory) DebugInfo() map[string][]string {
 	return map[string][]string{}
 }
 
+var (
+	version_regexp_string = "(\\d+)\\.(\\d+)\\.(\\d+)"
```
Comment (timstclair) [inline container/docker/factory.go]: Maybe change this to a raw string so you don't need to escape the backslash?

File: container/docker/factory.go (hunk 1 above)
Comment (jimmidyson) [inline container/docker/factory.go]: Putting it in the function means the RE gets wastefully compiled every time the function is called. In this case probably not called more than once, but normally best to move it outside so trying to keep it consistent.

File: container/docker/factory.go (hunk 2 above)
Comment (jimmidyson) [inline container/docker/factory.go]: Good catch,

File: container/libcontainer/helpers.go [hunk 3]
```diff
@@ -224,16 +221,14 @@ func scanTcpStats(tcpStatsFile string) (info.TcpStat, error) {
...
-		matched := tcpStatLineRE.MatchString(line)
-
-		if matched {
-			state := strings.Fields(line)
-			//#file header tcp state is the 4 filed:
-			//sl local_address rem_address st tx_queue rx_queue tr tm->when retrnsmt  uid timeout inode
-			tcpStateMap[state[3]]++
+
+		state := strings.Fields(line)
+		//#file header tcp state is the 4th field:
```
Comment (timstclair) [inline container/libcontainer/helpers.go]: (nit) space after //

Comment (timstclair) [inline container/libcontainer/helpers.go]: Also this comment reads a little weird. Maybe rephrase to:
```
// TCP state is the 4th field.
// Format: sl local_address rem_address st tx_queue rx_queue tr tm->when retrnsmt  uid timeout inode
```

File: container/libcontainer/helpers.go [hunk 4]
```diff
@@ -224,16 +221,14 @@ func scanTcpStats(tcpStatsFile string) (info.TcpStat, error) {
...
+		//sl local_address rem_address st tx_queue rx_queue tr tm->when retrnsmt  uid timeout inode
+		tcpState := state[3]
+		if _, ok := tcpStateMap[tcpState]; ok {
```
Comment (timstclair) [inline container/libcontainer/helpers.go]: If this is not OK it means you encountered an unexpected state. I think it's worth logging an error or warning if that happens (although that could generate a LOT of log spam if the file is corrupted, so it might be good to rate limit to a single log entry per call)

File: fs/fs.go [hunk 5]
```diff
@@ -35,7 +35,7 @@ import (
 	"github.com/golang/glog"
 )
 
-var partitionRegex = regexp.MustCompile("^(:?(:?s|xv)d[a-z]+\\d*|dm-\\d+)$")
+var partitionRegex = regexp.MustCompile(`^(:?(:?s|xv)d[a-z]+\d*|dm-\d+)$`)
```
Comment (timstclair) [inline fs/fs.go]: Does this regex optionally start with `::`, or is that supposed to be a non-capturing group? If the latter, it's backwards:
`(?:re)`

Also would prefer to move this closer to where it's used.

File: container/docker/factory.go (hunk 1 above)
Comment (timstclair) [inline container/docker/factory.go]: Ah, makes sense. Maybe move the declaration to just above the function in that case? Also, consider making the string const.

File: metrics/prometheus_test.go [hunk 6]
```diff
@@ -152,6 +152,11 @@ func (p testSubcontainersInfoProvider) SubcontainersInfo(string, *info.Container
 	}, nil
 }
 
+var (
+	includeRe = regexp.MustCompile("^(?:(?:# HELP |# TYPE)container_|cadvisor_version_info{)")
```
Comment (timstclair) [inline metrics/prometheus_test.go]: Does this compile correctly? I would expect that `{` needs to be escaped.

File: utils/machine/machine.go [hunk 7]
```diff
@@ -34,12 +34,18 @@ import (
...
 // The utils/machine package contains functions that extract machine-level specs.
 
-var cpuRegExp = regexp.MustCompile("processor\\t*: +([0-9]+)")
-var coreRegExp = regexp.MustCompile("core id\\t*: +([0-9]+)")
-var nodeRegExp = regexp.MustCompile("physical id\\t*: +([0-9]+)")
-var CpuClockSpeedMHz = regexp.MustCompile("cpu MHz\\t*: +([0-9]+.[0-9]+)")
-var memoryCapacityRegexp = regexp.MustCompile("MemTotal: *([0-9]+) kB")
-var swapCapacityRegexp = regexp.MustCompile("SwapTotal: *([0-9]+) kB")
+var (
+	cpuRegExp            = regexp.MustCompile("processor\\t*: +([0-9]+)")
```
Comment (timstclair) [inline utils/machine/machine.go]: (nit) raw strings

Comment (timstclair) [inline utils/machine/machine.go]: I wonder if ` +` should be `\s+`?

Comment (timstclair) [inline utils/machine/machine.go]: Actually, same with `\t` (use `\s`)? Might be more robust.

Same below for all these.

File: utils/machine/machine.go [hunk 8]
```diff
@@ -34,12 +34,18 @@ import (
...
+	coreRegExp           = regexp.MustCompile("core id\\t*: +([0-9]+)")
+	nodeRegExp           = regexp.MustCompile("physical id\\t*: +([0-9]+)")
+	CpuClockSpeedMHz     = regexp.MustCompile("cpu MHz\\t*: +([0-9]+.[0-9]+)")
```
Comment (timstclair) [inline utils/machine/machine.go]: escape decimal? `\.`

File: utils/machine/machine.go [hunk 9]
```diff
@@ -34,12 +34,18 @@ import (
...
+	memoryCapacityRegexp = regexp.MustCompile("MemTotal: *([0-9]+) kB")
+	swapCapacityRegexp   = regexp.MustCompile("SwapTotal: *([0-9]+) kB")
+
+	cpuClockSpeedMHz = regexp.MustCompile("clock\\t*: +([0-9]+.[0-9]+)MHz")
```
Comment (timstclair) [inline utils/machine/machine.go]: here too: `\.`

Comment (timstclair) [inline utils/machine/machine.go]: Unused? Did you mean to remove the lines where this was used below? Also, consider differentiating the variable name more from `CpuClockSpeedMHz`

File: utils/oomparser/oomparser.go [hunk 10]
```diff
@@ -29,12 +29,11 @@ import (
...
-var containerRegexp *regexp.Regexp = regexp.MustCompile(
-	`Task in (.*) killed as a result of limit of (.*)`)
-var lastLineRegexp *regexp.Regexp = regexp.MustCompile(
-	`(^[A-Z]{1}[a-z]{2} .*[0-9]{1,2} [0-9]{1,2}:[0-9]{2}:[0-9]{2}) .* Killed process ([0-9]+) \(([0-9A-Za-z_]+)\)`)
-var firstLineRegexp *regexp.Regexp = regexp.MustCompile(
-	`invoked oom-killer:`)
+var (
+	containerRegexp *regexp.Regexp = regexp.MustCompile(`Task in (.*) killed as a result of limit of (.*)`)
+	lastLineRegexp  *regexp.Regexp = regexp.MustCompile(`(^[A-Z]{1}[a-z]{2} .*[0-9]{1,2} [0-9]{1,2}:[0-9]{2}:[0-9]{2}) .* Killed process ([0-9]+) \(([0-9A-Za-z_]+)\)`)
+	firstLineRegexp *regexp.Regexp = regexp.MustCompile(`invoked oom-killer:`)
```
Comment (timstclair) [inline utils/oomparser/oomparser.go]: Why are these pointers?

File: utils/oomparser/oomparser.go [hunk 11]
```diff
@@ -29,12 +29,11 @@ import (
...
 
-var containerRegexp *regexp.Regexp = regexp.MustCompile(
-	`Task in (.*) killed as a result of limit of (.*)`)
-var lastLineRegexp *regexp.Regexp = regexp.MustCompile(
-	`(^[A-Z]{1}[a-z]{2} .*[0-9]{1,2} [0-9]{1,2}:[0-9]{2}:[0-9]{2}) .* Killed process ([0-9]+) \(([0-9A-Za-z_]+)\)`)
-var firstLineRegexp *regexp.Regexp = regexp.MustCompile(
-	`invoked oom-killer:`)
+var (
+	containerRegexp *regexp.Regexp = regexp.MustCompile(`Task in (.*) killed as a result of limit of (.*)`)
+	lastLineRegexp  *regexp.Regexp = regexp.MustCompile(`(^[A-Z]{1}[a-z]{2} .*[0-9]{1,2} [0-9]{1,2}:[0-9]{2}:[0-9]{2}) .* Killed process ([0-9]+) \(([0-9A-Za-z_]+)\)`)
```
Comment (timstclair) [inline utils/oomparser/oomparser.go]: `{1}` unnecessary

(optional) `[0-9A-Za-z_]` can be replaced with `\w`, although I'm not sure that's more readable.

**Comment by jimmidyson** [comment]
@timstclair Thanks for a thorough code review - this is making it a much more worthwhile endeavour than it would have otherwise been :)

File: utils/sysinfo/sysinfo.go [hunk 12]
```diff
@@ -65,18 +65,14 @@ func GetBlockDeviceInfo(sysfs sysfs.SysFs) (map[string]info.DiskInfo, error) {
...
-		} else {
-			matches := schedulerRegExp.FindSubmatch([]byte(sched))
-			if len(matches) < 2 {
-				sched = "none"
-			} else {
-				sched = string(matches[1])
+		disk_info.Scheduler = "none"
+		blkSched, err := sysfs.GetBlockDeviceScheduler(name)
+		if err == nil {
+			matches := schedulerRegExp.FindSubmatch([]byte(blkSched))
```
Comment (timstclair) [inline utils/sysinfo/sysinfo.go]: This regex should use a raw string too.

File: container/libcontainer/helpers.go (hunk 4 above)
Comment (jimmidyson) [inline container/libcontainer/helpers.go]: Probably best to return an error - can you trust stats from a corrupt file?

Comment (timstclair) [inline container/libcontainer/helpers.go]: Yeah, makes sense.

File: utils/machine/machine.go (hunk 9 above)
Comment (jimmidyson) [inline utils/machine/machine.go]: Seems as though it was always unused... removing.

Comment (jimmidyson) [inline utils/machine/machine.go]: Ah hang on messed up removing stuff below. Will add it back in.

Comment (timstclair) [inline utils/machine/machine.go]: Actually, maybe just combine this with `CpuClockSpeedMHz`? Something like:
```
`(?:cpu MHz|clock)\t*: +([0-9]+\.[0-9]+)(?:MHz)?`
```

**Comment by timstclair** [comment]
> @timstclair Thanks for a thorough code review - this is making it a much more worthwhile endeavour than it would have otherwise been :)

No problem! Thanks for the cleanup!

File: utils/machine/machine.go (hunk 9 above)
Comment (jimmidyson) [inline utils/machine/machine.go]: Great minds...

**Comment by jimmidyson** [comment]
All comments addressed. Good review!

**Comment by jimmidyson** [comment]
As an observation there is a lot we could do to lower resource utilisation. Regexps are hungry & need to be used judiciously. They're overused here where we could probably do some cheaper string parsing.

**Comment by jimmidyson** [comment]
Integration tests pass, so it must be good right? :stuck_out_tongue_winking_eye:

File: utils/machine/machine.go [hunk 13]
```diff
@@ -157,6 +160,7 @@ func GetTopology(sysFs sysfs.SysFs, cpuinfo string) ([]info.Node, int, error) {
 				lastNode = -1
 			}
 			lastThread = thread
+			continue
```
Comment (timstclair) [inline utils/machine/machine.go]: This whole loop could use some cleanup. I think it would be much more efficient and readable to switch on `strings.TrimSpace(line[:strings.Index(line, ":")-1])`

That's sort of a separate change, so up to you where / whether you want to address it.

**Comment by timstclair** [comment]
LGTM (please squash)

> Integration tests pass, so it must be good right?

Of course! I always like to have some simple unit tests for regexes, but I won't force you to add them if you aren't up for it...

> As an observation there is a lot we could do to lower resource utilisation. Regexps are hungry & need to be used judiciously. They're overused here where we could probably do some cheaper string parsing.

Agreed. Although I don't think it's worth spending too much time on this without first doing some profiling to identify the places it really matters.

File: utils/sysinfo/sysinfo.go (hunk 12 above)
Comment (jimmidyson) [inline utils/sysinfo/sysinfo.go]: Done.

**Comment by jimmidyson** [comment]
Yeah we need to do more profiling of cadvisor in general. There have been quite a few performance issues when embedded in the kubelet & we need to identify areas for improvement. Regexes were a recurrent theme in those perf problems, hence me looking at them here. I think with the work we've done recently in cadvisor this is improved significantly, but further work needs to be done to keep cadvisor's footprint as low as possible.

File: utils/machine/machine.go (hunk 13 above)
Comment (jimmidyson) [inline utils/machine/machine.go]: Agreed. Not going to do it here, luckily AFAICT this is only called once on start of cadvisor so efficiency isn't paramount. Would like to improve it though, for vanity more than anything else :-b

**Comment by jimmidyson** [comment]
retest this please

**Comment by jimmidyson** [comment]
Self-merging after LGTM.
